### DISCUSSION TITLE
Spatial queries

### DISCUSSION THREAD

**Discussion by sebastianfrey** [ngageoint/geopackage-java#56] [Q&A] [answered]
Is it possible to query for features using spatial conditions?

**Comment by bosborn** [comment, marked as answer]
Something like this...

```
FeatureIndexManager indexManager = new FeatureIndexManager(
	geoPackage, featureTable);
if(!indexManager.isIndexed()) {
    indexManager.setIndexLocation(FeatureIndexType.RTREE);
    indexManager.index();
}
FeatureIndexResults results = indexManager.query(/* args including bounding box or envelope */);
for (FeatureRow featureRow : results) {
    // ...
}
results.close();
indexManager.close();
```

**Reply by sebastianfrey** [reply]
Thank you for your quick response and pointing me to the right API.

Now I have another question, after looking at the `FeatureIndexManger` API docs.

While the `FeatureDao` has built in support for paginated queries leveraging `#queryForChunk`, it seems there is no similar public API exposed by `FeatureIndexManager`.

What would be the appropriate approach for this? Or are paginated queries currently not possible, when using spatial queries.

**Reply by sebastianfrey** [reply]
@bosborn 

Ok, I got the paginated query working.

```
FeatureIndexManager indexManager = null;
FeatureIndexResults results = null;

try (GeoPackage geoPackage = GeoPackageManager.open("pot_WEG_Luebz_linie.gpkg")) {
  indexManager = new FeatureIndexManager(geoPackage, "pot_WEG_Luebz_linie");
  if (!indexManager.isIndexed()) {
    indexManager.setIndexLocation(FeatureIndexType.RTREE);
    indexManager.index();
  }

  Double minX = 33302120.0;
  Double minY = 5923548.0;
  Double maxX = 33303206.0;
  Double maxY = 5925394.0;

  RTreeIndexTableDao rTreeIndexTableDao = indexManager.getRTreeIndexTableDao();

  StringBuilder whereBuilder = new StringBuilder();

  whereBuilder.append(rTreeIndexTableDao.buildWhere(RTreeIndexExtension.COLUMN_MIN_X, minX, "<="));
  whereBuilder.append(" AND ");
  whereBuilder.append(rTreeIndexTableDao.buildWhere(RTreeIndexExtension.COLUMN_MIN_Y, minY, "<="));
  whereBuilder.append(" AND ");
  whereBuilder.append(rTreeIndexTableDao.buildWhere(RTreeIndexExtension.COLUMN_MAX_X, maxX, ">="));
  whereBuilder.append(" AND ");
  whereBuilder.append(rTreeIndexTableDao.buildWhere(RTreeIndexExtension.COLUMN_MAX_Y, maxY, ">="));

  String whereBounds = whereBuilder.toString();
FeatureIndexManager
  Double tolerance = rTreeIndexTableDao.getTolerance();

  minX -= tolerance;
  maxX += tolerance;
  minY -= tolerance;
  maxY += tolerance;

  String[] whereArgs = rTreeIndexTableDao.buildWhereArgs(new Object[] { maxX, maxY, minX, minY });

  String where = rTreeIndexTableDao.getFeatureDao().buildWhereIn(rTreeIndexTableDao.queryIdsSQL(whereBounds), null);

  FeatureResultSet featureResultSet = rTreeIndexTableDao.getFeatureDao().queryForChunk(where, whereArgs, "fid", 5, 0);

  int count = 0;
  try {
    while (featureResultSet.moveToNext()) {
      FeatureRow featureRow = featureResultSet.getRow();
      GeoPackageGeometryData geometryData = featureRow.getGeometry();
      if (geometryData != null && !geometryData.isEmpty()) {
        count++;
      }`
    }
  } finally {
    featureResultSet.close();
  }
  System.out.println("count=" + count);
} finally {
  if (results != null) {
    results.close();
  }
  if (indexManager != null) {
    indexManager.close();
  }
}
```

The above solution is far from perfect, since I had to rely on some internals (`RTreeIndexTableDao#buildWhere`, `RTreeIndexTableDao#buildWhereArgs`)

I think, it would be nice, when the `FeatureIndexManager` would have built in support for paginated queries.

**Reply by bosborn** [reply]
You are correct that the methods had not been propagated out to the index manager due to the nature of the "query in" calls.

I [added](https://github.com/ngageoint/geopackage-java/commit/f2cddf9d78a9b4133084a98c772743d3b5c5fb32) and have it working for the FeatureIndexManager (still working on the individual index types).  I'll have it released in the next version.

**Reply by sebastianfrey** [reply]
@bosborn Wow! That's awesome. Didn't expected such a fast response and nice solution. Big thumbs up! 👍

**Reply by bosborn** [reply]
Updates released in [version 6.2.0](https://github.com/ngageoint/geopackage-java/releases/tag/6.2.0).  Includes the query chunk methods in index manager & individual types, [automated paginated query helpers](https://github.com/ngageoint/geopackage-java/blob/73f7372718268d766a6f1a112e865a5270fafee1/src/test/java/mil/nga/geopackage/ReadmeTest.java#L192), and [iterable row results](https://github.com/ngageoint/geopackage-java/blob/73f7372718268d766a6f1a112e865a5270fafee1/src/test/java/mil/nga/geopackage/ReadmeTest.java#L131).
